extra/lrshift~/lrshift~.c

```diff
@@ -7,60 +7,61 @@ static t_class *lrshift_tilde_class;
 typedef struct _lrshift_tilde
 {
     t_object x_obj;
-    int x_n;
-    t_float x_f;
+    t_float  x_n;
+    t_float  x_f;
 } t_lrshift_tilde;
 
-static t_int *leftshift_perform(t_int *w)
+static t_int *lrshift_perform(t_int *w)
 {
-    t_sample *in = (t_sample *)(w[1]);
-    t_sample *out= (t_sample *)(w[2]);
-    int n = (int)(w[3]);
-    int shift = (int)(w[4]);
-    in += shift;
-    n -= shift;
-    while (n--)
-        *out++ = *in++;
-    while (shift--)
-        *out++ = 0;
-    return (w+5);
+    t_lrshift_tilde *x = (t_lrshift_tilde *) w[1];
+    t_sample *in = (t_sample *)(w[2]);
+    t_sample *out= (t_sample *)(w[3]);
+    int n = (int)(w[4]);
+    
+    //test for abs(shift) >= n and limit shift between -(n-1) and n
+    //when shift equals n, only the "*out++ = 0.0f" loop runs
+    int shift = (int)x->x_n;
+    shift = (((shift<0)?-1*shift:shift) < n)?shift:n;
```

Comment:
Pd uses spaces before and after operators.

Reply:
This is way too complicated. Later you have to branch over the value of `shift` anyway. So why not simply do it like this:
```
if (shift > 0) /* left shift */
{
    /* validate shift value, set pointers and perform */
}
if (shift < 0) /* right shift */
{
    /* validate shift value, set pointers and perform */
}
else /* no shift */
{
    /* just copy */
}
```

Reply:
Thanks very much for your feedback and advice.  There's a lot of room to simplify lrshift~.c and I think it should be done.

I wrote it the way I did because it was symmetric and short.  Some of the conditional logic is contained in the fact that a while/for loop may run 0 times.  
If shift=0, the result can be produced by either block.  Only the 1st loop in each block runs.
If abs(shift) > n-1, the input is never read.  Only the 2nd loop in each block runs.  So, these are trivial cases.

I'm slightly intimidated by the idea of optimizing lrshift~.c at the moment.  I'm reading d_arithmetic.c for reference while I'm working on it, and many routines there are written with "perform" and "perf8" perform routines.  I could emulate the style, but I am not a sophisticated programmer.

I'm in favor of having "code that speaks for itself." I have these places I left the comments in for the 1st draft, I hope it helps you read it quickly.  I can remove those comments as they become un-needed.  Maybe we can settle on a version of it that's clear enough it doesn't need them.

lrshift~ really ought to be a dirt simple and nearly optimal piece of code.  The specification is really, really simple: turn a floatarg into the integer number of samples to shift while copying one array to another.  This is a really basic array operation.  

The motivating case is in abstractions that have variable block sizes, lrshift~ should be able to change its parameters along with changes to the blocksize using messages.  I'm in favor of being able to set the number as default and also update it from float messages later.  

I've tested this code (as shift~.c external) and produced identical output as lrshift~.c does.  I'm learning the process of making a pull request for the first time--do you want to see a test patch or do any other tests with it next?

I was considering the number of conditional statements being evaluated on each block as well.  I can save one conditional per tick by moving the validation inside the if-statement blocks and it's less complicated.  I've made the change there.  I also noticed I could write the second for loop as a while loop with the same count when shift is a negative int.  

I think I addressed your style notes.  Thanks for helping me with those.
Please take another look at my perform routine again, and see if you think it should have more changes.

Reply:
The code is now much more readable. Thanks!